File: apps/native/modules/shell-executor/ios/ShellExecutorModule.swift
import ExpoModulesCore

public class ShellExecutorModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ShellExecutor")

    AsyncFunction("executeCommand") { (command: String) in
      return try await executeShellCommand(command: command)
    }
  }

  private func executeShellCommand(command: String) async throws -> [String: Any] {
    let task = Process()
    task.launchPath = "/bin/sh"
    task.arguments = ["-c", command]

    let outputPipe = Pipe()
    let errorPipe = Pipe()

    task.standardOutput = outputPipe
    task.standardError = errorPipe

    do {
      try task.run()
      task.waitUntilExit()

      let outputData = outputPipe.fileHandleForReading.readDataToEndOfFile()
      let errorData = errorPipe.fileHandleForReading.readDataToEndOfFile()

      let stdout = String(data: outputData, encoding: .utf8) ?? ""
      let stderr = String(data: errorData, encoding: .utf8) ?? ""

      return [
        "stdout": stdout,
        "stderr": stderr,
        "exitCode": Int(task.terminationStatus)
      ]
    } catch {
      return [
        "stdout": "",
        "stderr": error.localizedDescription,
        "exitCode": -1
      ]
    }
  }
}

File: apps/native/android/app/src/main/java/studio/nebaura/mote/ShellExecutorModule.kt
package studio.nebaura.mote

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader

class ShellExecutorModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ShellExecutor")

    AsyncFunction("executeCommand") { command: String ->
      return@AsyncFunction executeShellCommand(command)
    }
  }

  private suspend fun executeShellCommand(command: String): Map<String, Any> = withContext(Dispatchers.IO) {
    try {
      val process = Runtime.getRuntime().exec(arrayOf("/system/bin/sh", "-c", command))

      val stdout = BufferedReader(InputStreamReader(process.inputStream)).use { it.readText() }
      val stderr = BufferedReader(InputStreamReader(process.errorStream)).use { it.readText() }

      val exitCode = process.waitFor()

      mapOf(
        "stdout" to stdout,
        "stderr" to stderr,
        "exitCode" to exitCode
      )
    } catch (e: Exception) {
      mapOf(
        "stdout" to "",
        "stderr" to e.message.orEmpty(),
        "exitCode" to -1
      )
    }
  }
}
